Stricter enforcement of the "large space" heuristic

CodaFi opened the pull request:
The Space Engine includes a heuristic that attempted a combinatorics-
based check to see if a pattern covered an insufficient amount of cases.
In straight-line switches this avoided computing a subspace match that
would have been quite expensive computationally.  However, when
expressive patterns (like tuple patterns) are used, it can fool the
check because the covered space is much larger than the actual size of
the space due to pattern overlap that counting like this simply can't detect.

Instead weaken the heuristic to diagnose large spaces as "uncheckable" and suggest a `default` clause.  This requires disabling SILGen's warning for unreachable `default` clauses which is itself incomplete.

Resolves [SR-6316](https://bugs.swift.org/browse/SR-6316).

<strike>I'm concerned about the performance impact of this change, which is what drove us to use the heuristic in the first place.</strike>

CodaFi committed:
Stricter enforcement of the "large space" heuristic

The Space Engine includes a heuristic that attempted a combinatorics-
based check to see if a pattern covered an insufficient amount of cases.
In straight-line switches this avoided computing a subspace match that
would have been quite expensive computationally.  However, when
expressive patterns (like tuple patterns) are used, it can fool the
check because the covered space is much larger than the actual size of
the space due to pattern overlap that counting like this simply can't detect.

Instead, just do the right thing and perform a space subtraction after
the existing preconditions for the heuristic are satisfied.

Resolves SR-6316

CodaFi commented:
@swift-ci please smoke test

CodaFi commented:
(Tagging Joe because he was the last one I talked to about this)

jckarter commented:
@swift-ci please smoke test compiler performance

jckarter commented:
SGTM. @graydon might be able to help you interpret the compiler performance counters.

jckarter commented:
@swift-ci Please smoke test

jrose-apple commented:
The other thing we could do is err on the side of requiring the default. We could even phrase the diagnostic apologetically, i.e. "the compiler was unable to prove that all cases are handled".

CodaFi commented:
Trouble with that is SILOptimizer likes to pop unreachable diagnostics on default clauses.

jrose-apple commented:
Yeah…maybe we need a new kind of "unreachable" that says "yes, I know, but the frontend can't do better".

jckarter commented:
@jrose-apple That seems reasonable to me too. @CodaFi is there a reason we don't do that already?

jrose-apple commented:
Worth noting that Robert and I casually switched to talking about a different problem. This bug is about an uncovered switch that Sema thought was covered, and assuming "you need a default" would have fixed it. However, there are also covered switches that Sema might think are uncovered, but SIL diagnostics later prove to be covered. In this case we'd be back in a "damned if you do, damned if you don't" scenario.

jckarter commented:
The SIL diagnostic behavior pre-dates the space engine. I'd be fine with delegating the problem of exhaustiveness analysis to Sema and suppressing the diagnostic in SIL.

CodaFi commented:
Compiler perf benchmarks look like noise, which is good.  Gonna run some more tests locally to see if I can get it to regress on less common idioms.

CodaFi commented:
Also relevant is [SR-3278](https://bugs.swift.org/browse/SR-3278).  I will push another commit here that shows what removing the SIL diagnostic would look like.

rudkx commented:
@CodaFi Is this ready to go in?

rudkx commented:
@swift-ci Please test source compatibility

CodaFi commented:
I want to test it on one more set of inputs. I’ll merge by tomorrow and PR to 4.1 then.

rudkx commented:
@CodaFi Great, thanks!

CodaFi committed:
Defer analysis of large switch bodies

This presents a regression in diagnostic quality that is definitely
worth it not to lie to SILGen about whether a switch is covered or not.

At the same time, disable SIL’s unreachable diagnostic for ‘default’
clauses which would previously cause a warning to be emitted if the
default was proven to be unreachable.  This analysis is incomplete
anyways and can be done by Sema in the future if we desire.

CodaFi commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -3857,6 +3857,10 @@ WARNING(redundant_particular_literal_case,none,
 NOTE(redundant_particular_literal_case_here,none,
      "first occurrence of identical literal pattern is here", ())
 
+ERROR(cannot_prove_exhaustive_switch,none,
```
@jrose-apple Here's the new diagnostic

jckarter replied:
I think the diagnostic should also include a mention of the consequence of the "too large" condition so the user knows what action to take—they just have to provide a default, right?

jrose-apple replied:
This sounds good, along with what Joe said. We want the user to know immediately what to do next, even if there's some grumbling involved.

jckarter replied:
If it doesn't require too much SourceLoc gymnastics, can we have a fixit insert `default: <#fatalError()#>` or something like that?

CodaFi commented on lib/SILGen/SILGenPattern.cpp:
```diff
@@ -1031,7 +1031,8 @@ void PatternMatchEmission::emitDispatch(ClauseMatrix &clauses, ArgArray args,
         }
-        SGF.SGM.diagnose(Loc, diag::unreachable_case, isDefault);
+        if (!isDefault)
+          SGF.SGM.diagnose(Loc, diag::unreachable_case);
```
@jckarter Is this reasonable?

jckarter replied:
I don't recall the history here—why are we diagnosing this in SILGen?

jrose-apple replied:
Just checking, since I've been out of it: we do diagnose unreachable defaults in Sema when we can, right?

CodaFi replied:
As of now: no (and I don’t believe we should because of things like [SR-3278](https://bugs.swift.org/browse/SR-3278)).  `default` early-exits the current exhaustiveness algorithm.

jrose-apple replied:
Ah, right. I guess that's on me to put in once we switch to resilient builds with the exhaustive/non-exhaustive distinction on.

CodaFi commented:
@swift-ci please smoke test

CodaFi commented:
@swift-ci please test compiler performance

CodaFi commented:
Finally,

@swift-ci please test source compatibility

jckarter reviewed (approved):
LGTM.

CodaFi commented:
⛵️

rudkx commented:
@CodaFi Were you still planning on opening a 4.1 PR?

CodaFi commented:
@rudkx Done! #13438

rudkx commented:
Thanks!
